packages/monaco/src/browser/monaco-outline-contribution.ts

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable, inject } from "inversify";
+import SymbolInformation = monaco.modes.SymbolInformation;
+import SymbolKind = monaco.modes.SymbolKind;
+import { FrontendApplicationContribution, FrontendApplication, OpenerService, open, ITreeNode } from "@theia/core/lib/browser";
+import { MonacoTextModelService } from './monaco-text-model-service';
+import { Range, EditorManager, EditorWidget } from '@theia/editor/lib/browser';
+import DocumentSymbolProviderRegistry = monaco.modes.DocumentSymbolProviderRegistry;
+import CancellationTokenSource = monaco.cancellation.CancellationTokenSource;
+import { DisposableCollection } from "@theia/core";
+import { OutlineViewService, OutlineSymbolInformationNode } from "@theia/outline-view/lib/browser/outline-view-service";
+import URI from "@theia/core/lib/common/uri";
+
+@injectable()
+export class MonacoOutlineContribution implements FrontendApplicationContribution {
+
+    protected ids: string[] = [];
+    protected symbolList: NodeAndSymbol[] = [];
+    protected readonly toDispose = new DisposableCollection();
+    protected readonly outlineSymbolInformations: MonacoOutlineSymbolInformationNode[];
+
+    constructor(
+        @inject(OutlineViewService) protected readonly outlineViewManager: OutlineViewService,
+        @inject(EditorManager) protected readonly editorManager: EditorManager,
+        @inject(OpenerService) protected readonly openerService: OpenerService,
+        @inject(MonacoTextModelService) protected readonly textModelService: MonacoTextModelService) { }
+
+    onStart(app: FrontendApplication): void {
+        this.outlineViewManager.onDidChangeOpenState(async isOpen => {
+            this.updateOutline();
+        });
+        // let's skip the initial current Editor change event, as on reload it comes before the language sevrers have started,
+        // resulting in an empty outline.
+        setTimeout(() => {
+            this.editorManager.onCurrentEditorChanged(async (editor: EditorWidget) => {
+                this.updateOutlineForEditor(editor);
+            });
+        }, 1000);
+
+        DocumentSymbolProviderRegistry.onDidChange(event => {
+            this.updateOutline();
+        });
+
+        this.outlineViewManager.onDidSelect(async node => {
```

Comment:
you should try, but I think the code of `onDidSelect` handler could be identical to the code of `onDidOpen` handler